Frontend integration

rasdy opened the pull request:
This PR adds frontend support for GET-ing content for the intro, task assignment, and inital reflection pages.

Also, getInitReflectPage() in queries.js has been changed so that it additionally returns the body text and the prompt IDs for each prompt.

The other get**X**ReflectPage() functions in queries.js have _not_ been changed, but they should be.

The pages added to the scenarios in insert_example_data.sql are mis-ordered, so for the sake of the last mini-demo (at first I didn't realize the editor frontend couldn't add pages other than the intro), for scenario 1 I set the ORDER of the initial reflection to 3, and I changed the conversation page (previously ORDER 3) to a PLAIN page with text that says it's a task assignment.

So basically the pages are still out of order (other than scenario 1's first three pages).

rasdy committed:
Get frontend more ready for final mini-demo

rasdy committed:
Get ready for final mini-demo

 - Fix misordered pages example_data.sql (initial reflect and task)
 - Change initReflectPage response to return body text and prompt ids
 - Fix frontend get requests for intro, task, and init reflect page

rasdy committed:
Change 2nd page from conversation to task assignment in example_data

rlee287 commented on goon-universe/queries.js:
```diff
@@ -540,12 +540,18 @@ function getStakeholderDescriptions(scenarioID){
         }
-        callback(results.rows)
+        let response = {}
+        response.prompts = results.rows.map(row => ({
```
Pinging @miesoz to make sure he is aware of the format change here

kungoyal replied:
It looks good to me. I've tested it this way and it seems to work fine

miesoz replied:
Looked over it with Kunal, seems good.

rlee287 commented:
The changes to `queries.js` do a bit more manipulation with the query results than I did, so I will add that to my copy and do the same with the other related functions.

With relation to the example data, we got rid of the PLAIN page type in favor of having a body text, so a PLAIN page type is now incorrect. (This may not have been communicated properly to you so I'm repeating it here again).

I'm assuming that Goon Universe has been made aware of the change in callback format, so I'll propagate the changes to the other page functions. (If they do not want this then we can easily roll back that part.)

rlee287 committed:
Adjust getReflectPage functions based on frontend-integration stuffs

Cherry picked from the following commits:
- d4b6c79169e0e564dc3481eefb68ccc0ed326879 (source)
- 9aa89b3e752a9e06864137b5e0157cf550be9df5 (typofix comma)
- 587c6e78dadfadd6eea25934639f5e4d01a89a96 (typofix query)

rlee287 commented:
Everything on `queries.js` should be properly fixed up now.

kungoyal commented on goon-universe/queries.js:
```diff
@@ -539,33 +539,52 @@ function getStakeholderDescriptions(scenarioID){
+            text: row.prompt,
+            id: row.prompt_num
+        }))
+        response.body_text = results.rows[0].body_text
```
The code works fine in most cases. However, when a scenario id that does not exist in the database is passed, the code crashes with a 'cannot read property body_text of undefined'. This needs to fixed. I think a simple if(results.length == 0) check will solve the problem

kungoyal replied:
Returning(callback) an empty object if that condition is true will be the best thing to do since it will work with the API without having to change anything

rlee287 committed:
For getBlahReflectPage, callback on empty object if SQL returns nothing

(cherry picked from commit b8eda81ef7394c56f105996955f7fad2b60289d5)

rlee287 committed:
results.rows.length instead of results.length

rlee287 commented:
@kungoyal When you call `get*ReflectPage`, your callback function needs to change: you need to replace `result.length == 0` with something that uses the actual object that gets returned (an empty `{}` if the ID was invalid or something with `body_text` and `prompts` fields if the ID was valid).

rlee287 reviewed (changes requested):
Not ready to merge; see comment I just left

rlee287 committed:
Ensure that promptNum is passed in when adding in reflection responses

(cherry picked from commit e4b9997cf1eaa88898bc9404a95bf40369db5103)

rlee287 committed:
Give all reflect responses a promptNum arg (and not just init)

(cherry picked from commit 5c8f0529ad35ef9d2fab63fd68ed66cb8b7bc597)

kungoyal committed:
Changed error condition for getXReflectPage calls

rlee287 reviewed (dismissed):
Seems like it is fixed to me; but @rasdy please test locally before merging this in

kungoyal committed:
Fixed ON CONFLICT error

rlee287 commented:
Next steps from integration meeting today:

- Implement for loop to loop over responses when inserting/getting
- Merge this in
- Merge in #201 after this PR

kungoyal committed:
addInitReflectResponse is now called multiple times

abhiz123 committed:
Updated middle and final reflection to call multiple times

rasdy commented:
Seems to be mostly working, but I found a couple issues.

 - If the student or scenario ID doesn't exist in the db, a PUT initialReflection request results in an unhandled promise rejection from `addReflectResponse()`
 - The response I get from calling `GET /scenarios/initalReflection/response` with `studentID: 2, scenarioID: 1` is confusing. I get prompts 1 and 2 for scenario 1, but I also get prompt 1 for scenario 2?
Calling `PUT /scenarios/initialReflection` with `studentID: 2, scenarioID: 1, data: {1: "new response"}`, according to `GET /scenarios/initalReflection/response`, overwrites prompt 2 instead of 1.
 - Calling `GET /scenarios/initalReflection/response` with the wrong studentID or scenarioID results in a 200 OK response containing an empty list

rlee287 commented:
For the first point: in `addXReflectResponse`, what should we put in the callback, and how should we handle callbacks when the student or scenario doesn't exist? (We have something right now but I'm asking again to make sure we are all on the same page here.)

rlee287 committed:
Reflection response return should filter by submissions.scenario_id

rasdy committed:
Set up frontend initReflect page for response submission

rasdy committed:
Merge branch 'frontend-integration' of github.com:david-fisher/320-F20-Track-II into frontend-integration

miesoz reviewed (commented):
Did not mean to approve, I just saw the merge conflicts.

rlee287 reviewed (changes requested):
Merge conflicts are not a reason to deny approval for a PR, as those happen periodically (and we can simply talk out how to resolve those when they occur.) However, there are other issues remaining that @rasdy pointed out earlier, so I'll review this as "request changes" until those get fixed.

rlee287 committed:
Callback empty string for addReflectionResponse upon nonexistent ID

rlee287 commented:
Echoing messages from Track2 slack channel:

- First item should be fixed now (following communication with Goon Universe)
- Second item GET works correctly now; unable to reproduce the PUT issue
- Third item: code that handles this was explicitly commented out for an unknown reason

kungoyal committed:
Added null check for getXResponse

rlee287 committed:
getReflectResponse helper func+checking if IDs exist first

rlee287 reviewed (dismissed):
Bullet points should be fixed now

rasdy committed:
Merge branch 'master' into frontend-integration
